File: frontend/src/services/projects.ts
import { apiClient } from "@/services/apiClient";
import type {
  PaginatedProjects,
  Project,
  ProjectFilters,
  ProjectFormValues,
  ProjectStats,
} from "@/types/project";

export async function fetchProjectStats(): Promise<ProjectStats> {
  const { data } = await apiClient.get<ProjectStats>("/admin/projects/stats");
  return data;
}

export async function fetchProjects(filters: ProjectFilters = {}): Promise<PaginatedProjects> {
  const { data } = await apiClient.get<PaginatedProjects>("/admin/projects", {
    params: {
      status: filters.status || undefined,
      q: filters.q || undefined,
      page: filters.page ?? 1,
      page_size: filters.page_size ?? 20,
    },
  });
  return data;
}

export async function fetchProject(id: number): Promise<Project> {
  const { data } = await apiClient.get<Project>(`/admin/projects/${id}`);
  return data;
}

function toFormData(values: ProjectFormValues, image?: File | null): FormData {
  const form = new FormData();
  form.append("slug", values.slug);
  form.append("title", values.title);
  form.append("industry", values.industry);
  form.append("service", values.service);
  form.append("location", values.location);
  form.append("live_url", values.live_url);
  form.append("image_url", values.image_url || "");
  form.append("image_alt", values.image_alt || "");
  form.append("status", values.status);
  form.append("featured", String(values.featured));
  form.append("sort_order", String(values.sort_order));
  if (image) form.append("image", image);
  return form;
}

export async function createProject(values: ProjectFormValues, image?: File | null): Promise<Project> {
  const { data } = await apiClient.post<Project>("/admin/projects", toFormData(values, image), {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return data;
}

export async function updateProject(
  id: number,
  values: ProjectFormValues,
  image?: File | null,
): Promise<Project> {
  const { data } = await apiClient.patch<Project>(
    `/admin/projects/${id}`,
    toFormData(values, image),
    { headers: { "Content-Type": "multipart/form-data" } },
  );
  return data;
}

export async function deleteProject(id: number): Promise<void> {
  await apiClient.delete(`/admin/projects/${id}`);
}

export async function fetchPublishedProjects(pageSize = 50): Promise<Project[]> {
  const { data } = await apiClient.get<PaginatedProjects>("/public/projects", {
    params: { page: 1, page_size: pageSize },
  });
  return data.items;
}

File: backend/app/api/v1/admin/projects.py
from fastapi import APIRouter, Depends, File, Form, Query, UploadFile
from sqlalchemy.orm import Session

from app.api.deps import get_db, require_role
from app.models.user import User
from app.schemas.project import PaginatedProjects, ProjectCreate, ProjectOut, ProjectStats, ProjectUpdate
from app.services.project_service import ProjectService

router = APIRouter(tags=["admin-projects"])

_view_roles = require_role("admin", "editor", "viewer")
_edit_roles = require_role("admin", "editor")


@router.get("/projects/stats", response_model=ProjectStats)
def project_stats(
    db: Session = Depends(get_db),
    _: User = Depends(_view_roles),
) -> ProjectStats:
    return ProjectService(db).get_stats()


@router.get("/projects", response_model=PaginatedProjects)
def list_projects(
    status: str | None = Query(default=None),
    q: str | None = Query(default=None, max_length=200),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: Session = Depends(get_db),
    _: User = Depends(_view_roles),
) -> PaginatedProjects:
    return ProjectService(db).list_projects(
        status=status,
        q=q,
        page=page,
        page_size=page_size,
    )


@router.get("/projects/{project_id}", response_model=ProjectOut)
def get_project(
    project_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(_view_roles),
) -> ProjectOut:
    return ProjectService(db).get_project(project_id)


@router.post("/projects", response_model=ProjectOut, status_code=201)
async def create_project(
    slug: str = Form(...),
    title: str = Form(...),
    industry: str = Form(...),
    service: str = Form(...),
    location: str = Form(...),
    live_url: str = Form(...),
    image_url: str = Form(default=""),
    image_alt: str = Form(default=""),
    status: str = Form(default="draft"),
    featured: bool = Form(default=False),
    sort_order: int = Form(default=0),
    image: UploadFile | None = File(default=None),
    db: Session = Depends(get_db),
    _: User = Depends(_edit_roles),
) -> ProjectOut:
    body = ProjectCreate(
        slug=slug,
        title=title,
        industry=industry,
        service=service,
        location=location,
        image_url=image_url or "pending",
        image_alt=image_alt,
        live_url=live_url,
        status=status,
        featured=featured,
        sort_order=sort_order,
    )
    if (not image or not image.filename) and (not image_url or image_url == "pending"):
        from fastapi import HTTPException

        raise HTTPException(status_code=400, detail="Provide an image file or image_url.")
    if image_url and image_url != "pending":
        body.image_url = image_url
    return ProjectService(db).create_project(body, image=image)


@router.patch("/projects/{project_id}", response_model=ProjectOut)
async def update_project(
    project_id: int,
    slug: str | None = Form(default=None),
    title: str | None = Form(default=None),
    industry: str | None = Form(default=None),
    service: str | None = Form(default=None),
    location: str | None = Form(default=None),
    live_url: str | None = Form(default=None),
    image_url: str | None = Form(default=None),
    image_alt: str | None = Form(default=None),
    status: str | None = Form(default=None),
    featured: bool | None = Form(default=None),
    sort_order: int | None = Form(default=None),
    image: UploadFile | None = File(default=None),
    db: Session = Depends(get_db),
    _: User = Depends(_edit_roles),
) -> ProjectOut:
    body = ProjectUpdate(
        slug=slug,
        title=title,
        industry=industry,
        service=service,
        location=location,
        live_url=live_url,
        image_url=image_url,
        image_alt=image_alt,
        status=status,
        featured=featured,
        sort_order=sort_order,
    )
    return ProjectService(db).update_project(project_id, body, image=image)


@router.delete("/projects/{project_id}", status_code=204)
def delete_project(
    project_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(_edit_roles),
) -> None:
    ProjectService(db).delete_project(project_id)
